_posts/2019-01-04-nextjs.md

```diff
@@ -0,0 +1,290 @@
+
+- Code splitting. For each page, only the necessary JavaScript code should be loaded, not the entire application. When moving to the next page, the missing code should be reloaded - without reloading the same modules, and without unnecessary modules.
+
+All these tasks are successfully solved by the Next.js library. This library is based on a very simple idea. It is proposed to introduce a new type of component - “page”, in which there is an async static method async getInitialProps ({req}). The “page” component is a normal React component. This type of component can be thought of as a new type in the series: “component”, “container”, “page”.
+
+## Sample Project
+
+Let's create a simple project with Next.js to demonstrate how it works. You will be surprised how little we need to do in order to start our actual application. Some complexity is hidden by the library and the rule convention over configuration is helping to keep code on our side really slim.
+
+Create a new directory for our sample project `mkdir sample-app && cd sample-app`. Generate an empty npm project here with `npm init -y`. Then install Next.js together with React dependencies: `npm i next react react-dom --save`.
+
+Now we just have to add a new npm run script to our `package.json`:
+
+```js
+  "scripts": {
+    "dev": "next"
+  }
+```
+
+Let's try: `npm run dev`. Unfortunately you'll see an error message in your terminal. But with  next.js, error messages are usually pretty good and tell you exactly what you have to do next:
+
+`Couldn't find a `pages` directory. Please create one under the project root`
+
+So go ahead, create a `pages` directory and rerun `npm run dev`. It will tell you that it is now running on port 3000 - so open localhost:3000 in your browser. It's rendering a 404 page however because we did not yet create any "page".
+
+<img src="/img/posts/2019-01-04-nextjs/404.png" alt="404 Not Found" class="centered" />
+
+Lets create an `index.js` file with the following content:
+
+`export default () => <div>Hello World</div>`
+
+If you refresh your browser you'll see our new awesome "Hello World" page. This sounds like very little but what actually happened under the hood could take you quite some time if you really do it from scratch:
+
+- we got a free routing layer that is based on the convention that every path maps to a file in the pages folder (`/` is index.js, `/hello` is `hello.js`)
+- our code is automatically transpiled - we can use `export` without any hassle
+- we automatically can use react and `jsx` without any webpack setup
+- if you change the text from `Hello World` to `Hello World!` it will automatically be refreshed in your browser
+- ...
+
+
+But lets focus on the "application" we want to build.
+
+## Showing hotel data
+
+We want to show data about an hotel. So let's create the necessary files for that:
```

Comment:
**a** hotel

Reply:
Oh crap. I always thought that `an hotel` is correct. Like `an hour`.  Turns out I was mislead because the `h` in `hotel` isn't silent. Thanks :)